### PR TITLE
[singlejar] cleanup port library

### PR DESCRIPTION
**Pull request by rongjiecomputer**
/cc @laszlocsomor

### CODE REVIEW

File: src/tools/singlejar/port.h [hunk 1]
```diff
@@ -19,13 +19,26 @@
...
 #include <fcntl.h>
 #include <io.h>
 #include <stddef.h>
 #include <stdio.h>
+#include <sys/stat.h>
 #include <sys/types.h>
 #include <time.h>
 
+#if defined(__APPLE__)
+typedef off_t off64_t;
```
Comment (laszlocsomor) [inline src/tools/singlejar/port.h]: Could you add a [`static_assert`](https://en.cppreference.com/w/cpp/language/static_assert) for the `sizeof(off_t)`, just to make sure it is indeed 8 bytes?

File: src/tools/singlejar/port.h [hunk 2]
```diff
@@ -36,7 +49,7 @@
 #define fstat _fstat64
 #endif  // _WIN64
 
-typedef ptrdiff ssize_t;
+typedef ptrdiff_t ssize_t;
```
Comment (laszlocsomor) [inline src/tools/singlejar/port.h]: WDYT about using [`std::ptrdiff_t`](https://en.cppreference.com/w/cpp/types/ptrdiff_t)? I'd ~hope~ expect the STL types are more likely to be the same on all platforms than the POSIX types.

Comment (rongjiecomputer) [reply src/tools/singlejar/port.h]: C++ `std::ptrdiff_t` is the same as C `ptrdiff_t`. It is implemented as `namespace std { using ::ptrdiff_t; }` in `<cstddef>`.

File: src/tools/singlejar/port.h [hunk 3]
```diff
@@ -19,13 +19,27 @@
...
 #include <time.h>
 
+#if defined(__APPLE__)
+typedef off_t off64_t;
+#elif defined(_WIN32)
+typedef __int64 off64_t;
+#endif
+static_assert(sizeof(off64_t) == 8, "File offset type must be 64-bit");
+
+#ifdef _WIN32
```
Comment (laszlocsomor) [inline src/tools/singlejar/port.h]: Wait, this one has no terminating `#endif`

File: src/tools/singlejar/output_jar.h [hunk 4]
```diff
@@ -15,6 +15,9 @@
 #ifndef SRC_TOOLS_SINGLEJAR_COMBINED_JAR_H_
 #define SRC_TOOLS_SINGLEJAR_COMBINED_JAR_H_
 
+// Must be included before anything else.
+#include "src/tools/singlejar/port.h"
```
Comment (laszlocsomor) [inline src/tools/singlejar/output_jar.h]: This file doesn't need the whole `port.h`, just the `off64_t` type definition. Could you split that out so including `port.h` (and its transitive closure of other header files) is unnecessary here?

File: src/tools/singlejar/port.h (hunk 3 above)
Comment (rongjiecomputer) [reply src/tools/singlejar/port.h]: https://github.com/rongjiecomputer/bazel/blob/257beec22b632c9535d0c9d6e4218f2dbf3c0593/src/tools/singlejar/port.h

- Line 18 is closed by line 23.
- Line 32 is closed by line 36.
- Line 39 is closed by line 67.
- In the third block, line 43 is closed by line 51.

I need to include `<sys/stat.h>` for all platforms to get `off_t` and `off64_t`.

Comment (laszlocsomor) [reply src/tools/singlejar/port.h]: Ah indeed, you added an `#endif` in line 23.

File: src/tools/singlejar/output_jar.h (hunk 4 above)
Comment (laszlocsomor) [reply src/tools/singlejar/output_jar.h]: > I need to include <sys/stat.h> for all platforms to get off_t and off64_t.

I suppose you meant that line as a reply to this comment of mine.
I think it's OK to include what's necessary to define `off64_t`, so `<sys/stat.h>` is fine, but not more. Including things like `<io.h>` on Windows and all the other system headers is too much -- they might pollute the preprocessor namespace and global namespace.

**Comment by laszlocsomor** [comment]
FYI, the test failures are related, e.g. https://source.cloud.google.com/results/invocations/61e22fcc-6413-4704-8cc3-f8c66cae32bb/targets

**Comment by rongjiecomputer** [comment]
`//src/test/shell/bazel:embedded_tools_deps_test` passed, but lots of other tests failed, likely not related to my change. I just rebased against master.

**Comment by rongjiecomputer** [comment]
> Is it necessary to have this inclusion above the angle-bracket-included library headers, or is it safe to move it below them (above the other quote-included files)?

As long as `port.h` is included before `<io.h>` and `<fnctl.h>`, it will be OK.

**Pull request closed without merging**